### ISSUE TITLE
ComputeAt with multiple different broadcast consumers fails to generate indices

### ISSUE THREAD

**Issue by naoyam** [csarofeen/pytorch#322]
## 🐛 Bug

Compiling this fusion fails at index computation.

```
  TensorView* tv0 = makeDummyTensor(1);
  fusion.addInput(tv0);

  TensorView* tv1 = mul(tv0, new Float(0.5));
  TensorView* tv2 = broadcast(tv1, {true, false});
  TensorView* tv3 = broadcast(tv1, {false, true});
  TensorView* tv4 = add(tv2, tv3);
  fusion.addOutput(tv4);

  tv1->computeAt(tv3, -1);
```

The error I got is:

```
C++ exception with description "index_map.find(kir_root_dom_i) != index_map.end() INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_compute.cpp":1060, please report a bug to PyTorch. Couldn't find root mapping for TV2 dim: 1 id: iS{T0.size[0]}
Exception raised from getConsumerIndex_impl at ../torch/csrc/jit/codegen/cuda/index_compute.cpp:1060 (most recent call first):
```

The full test code is available as testGPU_FusionComputeAtMultiBCast in https://github.com/naoyam/pytorch/blob/compute-at-bugs/test/cpp/jit/test_gpu.cpp.

**Comment by naoyam** [comment]
This also fails with a similar error message:

```
 auto tv0 = makeConcreteTensor({x, y});
  fusion.addInput(tv0);

  auto tv1 = add(tv0, new Float(1));

  auto tv2 = sum(tv1, {1});
  fusion.addOutput(tv2);
  auto tv3 = sum(tv1, {1});
  fusion.addOutput(tv3);

  tv1->computeAt(tv2, -1);
```

```
C++ exception with description "index_map.find(kir_root_dom_i) != index_map.end() INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_compute.cpp":786, please report a bug to PyTorch. Couldn't find root mapping for TV0 dim: 1 id: iS{200}
Exception raised from getGlobalProducerIndex at ../torch/csrc/jit/codegen/cuda/index_compute.cpp:786 (most recent call first):
```

See `testGPU_FusionComputeAtMultiReduction` test in https://github.com/naoyam/pytorch/blob/bcast-compute-at-bug/test/cpp/jit/test_gpu.cpp.

**Comment by naoyam** [comment]
> ## 🐛 Bug
> Compiling this fusion fails at index computation.
> 
> ```
>   TensorView* tv0 = makeDummyTensor(1);
>   fusion.addInput(tv0);
> 
>   TensorView* tv1 = mul(tv0, new Float(0.5));
>   TensorView* tv2 = broadcast(tv1, {true, false});
>   TensorView* tv3 = broadcast(tv1, {false, true});
>   TensorView* tv4 = add(tv2, tv3);
>   fusion.addOutput(tv4);
> 
>   tv1->computeAt(tv3, -1);
> ```
> 
> The error I got is:
> 
> ```
> C++ exception with description "index_map.find(kir_root_dom_i) != index_map.end() INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_compute.cpp":1060, please report a bug to PyTorch. Couldn't find root mapping for TV2 dim: 1 id: iS{T0.size[0]}
> Exception raised from getConsumerIndex_impl at ../torch/csrc/jit/codegen/cuda/index_compute.cpp:1060 (most recent call first):
> ```
> 
> The full test code is available as testGPU_FusionComputeAtMultiBCast in https://github.com/naoyam/pytorch/blob/compute-at-bugs/test/cpp/jit/test_gpu.cpp.

The tensors look as follows before the computeAt:

```
T1[ iS1{i1} ]
   = T0[ iS0{i1} ]
   * float(0.5);
T2[ bS2{1}, iS1{i1} ] = broadcast( T1[ iS1{i1} ] )
T3[ iS1{i1}, bS3{1} ] = broadcast( T1[ iS1{i1} ] )
T4[ iS5{i1}, iS4{i1} ]
   = T2[ bS2{1}, iS1{i1} ]
   + T3[ iS1{i1}, bS3{1} ];
```

The final loop structure after the computeAt should look as follows:

```
for x in i1
  T3[x] = T1[x]
  for y in i1
    T2[y] = T1[y]
    T4[x, y] = T2[y] + T3[x]
  end for
end for
```

However, the computeAt transforms the tensors as follows:

```
T1[ iS1{i1} ] compute_at( T3, 2 )
   = T0[ iS0{i1} ]
   * float(0.5);
T2[ iS1{i1}, bS2{1} ] compute_at( T4, 1 ) = broadcast( T1[ iS1{i1} ] compute_at( T3, 2 ) )
T3[ iS1{i1}, bS3{1} ] compute_at( T4, 1 ) = broadcast( T1[ iS1{i1} ] compute_at( T3, 2 ) )
T4[ iS5{i1}, iS4{i1} ]
   = T2[ iS1{i1}, bS2{1} ] compute_at( T4, 1 )
   + T3[ iS1{i1}, bS3{1} ] compute_at( T4, 1 );
```

Notice that the two dimensions of `T2` are reversed. That is done by `replayCasP`, which is run with `T2` as a consumer and `T1` as a producer at the producer position of 1.

```
Consumer: T2[ bS2{1}, iS1{i1} ]
Producer: T1[ iS1{i1} ] compute_at( T3, 2 )
Position: 1
```

After this `replayCasP`, `T2` is `T2[ iS1{i1}, bS2{1} ]` and the position is 1. This is because `replayCasP` tries to match the producer iterdom of `T1` with `T2`, which is `iS1`, so it is placed at the first position, and the computeAt position becomes 1. But the position should really be 2 since T2 should really be computed at position 2 of T4. 

In a nutshell, the root cause of this problem is that `replayCasP` of `T2` and `T1` reorders the two dimensions of `T2`.

**Comment by naoyam** [comment]
> This also fails with a similar error message:
> 
> ```
> auto tv0 = makeConcreteTensor({x, y});
> fusion.addInput(tv0);
> auto tv1 = add(tv0, new Float(1));
> auto tv2 = sum(tv1, {1});
> fusion.addOutput(tv2);
> auto tv3 = sum(tv1, {1});
> fusion.addOutput(tv3);
> 
> tv1->computeAt(tv2, -1);
> ```
> 
> ```
> C++ exception with description "index_map.find(kir_root_dom_i) != index_map.end() INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_compute.cpp":786, please report a bug to PyTorch. Couldn't find root mapping for TV0 dim: 1 id: iS{200}
> Exception raised from getGlobalProducerIndex at ../torch/csrc/jit/codegen/cuda/index_compute.cpp:786 (most recent call first):
> ```
> 
> Note that it works when the position of the computeAt is changed to -2.
> 
> See `testGPU_FusionComputeAtMultiReduction` test in https://github.com/naoyam/pytorch/blob/bcast-compute-at-bug/test/cpp/jit/test_gpu.cpp.

I believe this is a separate problem, so I created a new issue #329.

**Comment by naoyam** [comment]
This one actually needs recomputation of `tv1`, and since we don't do automatic recomputation (#139), there is no valid code possible for this fusion. So, this is actually not a bug, though we don't do a good job of explaining what the real problem is.

More specifically, the possible code that embeds `tv1` in the loop nest of `tv3` should look like as follows:

```
for x in I1
  for y in I1
    tv1_for_tv2 = tv0[y] * 0.5
    tv2 = tv1_for_tv1 
    tv1_for_tv3 = tv0[x] * 0.5      
    tv3 = tv1_for_tv3
    tv4[x, y] = tv2 + tv3
```

We could reorder `tv3`'s two dimensions so that they match with `tv2`'s dimensions, but then there would be a consistency problem with `tv4`. If `tv3` were reordered, that would mean `tv4` would also be reordered, but then that order would not match with `tv2`.

Until we can automatically recompute tensors as needed, what we do is to throw an error for problems like this. #327 should detect the inconsistent reordering. Ideally, we should throw more useful errors that can direct the user to understand what the real issue is.

**Comment by naoyam** [comment]
#423 detects this and throws an exception before doing transformations.
